### PR TITLE
Add Bazel support for C/C++ dependency chain

### PR DESCRIPTION
**Pull request by mobileink**
This PR adds Bazel support for building all C/C++ dependencies.  Tested on MacOS Catalina and Debian 9:

```
bazel test @libsnark//test
bazel test @libfqfft//test
bazel test @libff//test
bazel test @ate_pairing//test:CurveSNARK
bazel test @ate_pairing//test:CurveNoSNARK
bazel test @xbyak//test
```

For more info see doc/bazel.md.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -20,6 +20,13 @@ The best place to get started learning how to use the library are the annotated
 - [Election](examples/election/election_main.ml): shows how to use Snarky to verify an election was run honestly.
 - [Merkle update](examples/merkle_update/merkle_update.ml): a simple example updating a Merkle tree.
 
+### dependencies
+
+OPAM deps
+
+* ctyles 0.17.1
```
Comment (mrmr1993) [inline README.md:27]: I think this should be `ctypes`

File: doc/bazel.md [hunk 2]
```diff
@@ -78,7 +114,171 @@ build --verbose_failures
...
+[       OK ] gadgetLib2.Protoboard_multipackedWordAssignmentEqualsValue_R1P (0 ms)
+[ RUN      ] gadgetLib2.Protoboard_dualWordAssignmentEqualsValue_R1P
+[       OK ] gadgetLib2.Protoboard_dualWordAssignmentEqualsValue_R1P (1 ms)
+[----------] 4 tests from gadgetLib2 (2 ms total)
+
+[----------] Global test environment tear-down
+[==========] 4 tests from 1 test suite ran. (2 ms total)
+[  PASSED  ] 4 tests.
+
+That's on MacOS; the error msg is slightly different on Linux.
```
Comment (mrmr1993) [inline doc/bazel.md:179]: Nit: `message`

File: doc/bazel.md [hunk 3]
```diff
@@ -78,7 +114,171 @@ build --verbose_failures
...
+`--override_repository` parameter. This functions just like the
+`local_repository` rule, except that it only accepts absolute paths.
+You can avoid lots of typing by putting such parameters in a file
+named `user.bazelrc` located in the root directory of your main
+workspace, and putting `try-import user.bazelrc` in file `.bazelrc` in
+the same directory.
+
+Here is an example of how one might work on snarky and its
+depenencies. Create a working directory (let's call it snarky-dev),
+and clone snarky nd all the dependencies you want to work on into that
```
Comment (mrmr1993) [inline doc/bazel.md:237]: Nit: `and` (or maybe `with`)

File: doc/bazel.md [hunk 4]
```diff
@@ -171,74 +369,36 @@ For offline work you can prefetch remote external repos. See [Offline
...
-query --override_repository=xbyak=/PATH/TO/snarky-dev/xbyak
-
-```
-
-Now whenever you run `$ bazel build` or `$ bazel query` from
-`snarky-dev/snarky`, any changes you have made to your local
-dependency copies will be picked up. Once you have finished your edits
-and pushed them to the origin server, comment out these overrides to
-resume use of the repository rules in WORKSPACE.
+THe first three targets listed above build .a libs corresponding to
```
Comment (mrmr1993) [inline doc/bazel.md:394]: Nit: `The`

File: bzl/host/BUILD.bazel [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+
+config_setting(
+    name = "linux",
+    constraint_values = [
+        "@platforms//os:linux",
+    ]
+)
+
+config_setting(
+    name = "linux_arm",
```
Comment (mrmr1993) [inline bzl/host/BUILD.bazel:11]: I'm not sure that we support ARM at the moment (and don't have an ARM machine to test on). Have you tested on ARM? If not, I think we should remove this or comment it out, to make sure we're not giving an impression of ARM support where it doesn't necessarily exist.

File: examples/BUILD.bazel [hunk 6]
```diff
@@ -0,0 +1,40 @@
+alias(
```
Comment (mrmr1993) [inline examples/BUILD.bazel:1]: I'm a little uncomfortable with the libsnark examples being exposed so high up in the snarky repo. Can we move this down or remove these aliases?

**Comment by mobileink** [comment]
Cancelling, C++ no longer used.

**Pull request closed without merging**